File: codes/ruby/chapter09/answer_A68.rb
# Structure for edges used in maximum flow
class MaxFlowEdge
  attr_accessor :to, :cap, :rev

  def initialize(to, cap, rev)
    @to = to
    @cap = cap
    @rev = rev
  end
end

# Depth-first search
def dfs(pos, goal, f, g, used)
  return f if pos == goal # Arrived at the goal: The flow can be passed!

  # Explore
  used[pos] = true
  g[pos].each do |e|
    # Go only to vertices with a capacity of 1 or more and not yet visited
    if e.cap > 0 && !used[e.to]
      flow = dfs(e.to, goal, [f, e.cap].min, g, used)
      # If the flow can be passed, increase/decrease the capacity of the residual graph by flow
      if flow >= 1
        e.cap -= flow
        g[e.to][e.rev].cap += flow
        return flow
      end
    end
  end
  # Not found after exploring all edges...
  0
end

# Returns the total flow of the maximum flow from vertex s to vertex t (number of vertices N, list of edges)
def maxflow(n, s, t, edges)
  # Construct the initial residual graph
  g = Array.new(n + 1) { [] }
  edges.each do |a, b, c|
    g[a] << MaxFlowEdge.new(b, c, g[b].size)
    g[b] << MaxFlowEdge.new(a, 0, g[a].size - 1)
  end
  inf = 10**10
  total_flow = 0
  loop do
    used = Array.new(n + 1, false)
    f = dfs(s, t, inf, g, used)
    if f > 0
      total_flow += f
    else
      break # Finish operation when the flow can no longer be passed
    end
  end
  total_flow
end

# Input
n, m = gets.split.map(&:to_i)
edges = Array.new(m) { gets.split.map(&:to_i) }

# Find and output the answer
answer = maxflow(n, 1, n, edges)
puts answer

File: codes/go/chapter09/answer_A68.go
package main

import "fmt"

type FlowEdge struct {
	to  int
	cap int
	rev int
}

type MaximumFlow struct {
	n    int
	used []bool
	g    [][]FlowEdge
}

func newMaximumFlow(n int) *MaximumFlow {
	mf := MaximumFlow{
		n:    n,
		used: make([]bool, n+1),
		g:    make([][]FlowEdge, n+1),
	}

	for i := 0; i <= n; i++ {
		mf.g[i] = make([]FlowEdge, 0)
	}

	return &mf
}

func (mf *MaximumFlow) addEdge(a, b, c int) {
	mf.g[a] = append(mf.g[a], FlowEdge{b, c, len(mf.g[b])})
	mf.g[b] = append(mf.g[b], FlowEdge{a, 0, len(mf.g[a]) - 1})
}

func (mf *MaximumFlow) dfs(pos, goal, f int) int {
	if pos == goal {
		return f
	}

	mf.used[pos] = true
	for i, e := range mf.g[pos] {
		if e.cap == 0 || mf.used[e.to] {
			continue
		}
		flow := mf.dfs(e.to, goal, min(f, e.cap))
		if flow > 0 {
			mf.g[pos][i].cap -= flow
			mf.g[e.to][e.rev].cap += flow
			return flow
		}
	}
	return 0
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func (mf *MaximumFlow) maxFlow(s, t int) int {
	totalFlow := 0
	for {
		for i := 0; i <= mf.n; i++ {
			mf.used[i] = false
		}
		f := mf.dfs(s, t, int(^uint(0)>>1)) // Equivalent to Float64(INFINITY) in Ruby
		if f == 0 {
			break
		}
		totalFlow += f
	}
	return totalFlow
}

func main() {
	var n, m int
	fmt.Scanf("%d %d", &n, &m)
	z := newMaximumFlow(n)
	for i := 0; i < m; i++ {
		var a, b, c int
		fmt.Scanf("%d %d %d", &a, &b, &c)
		z.addEdge(a, b, c)
	}
	result := z.maxFlow(1, n)
	fmt.Println(result)
}
